Add documentation of how to use the Duende.IdentityModel.OidcClient.Extensions package to implement DPoP (#766)
DuendeSoftware/docs.duendesoftware.com: src/content/docs/identitymodel-oidcclient/advanced/dpop.mdx

---
title: Demonstrating Proof-of-Possession (DPoP)
description: Learn how to leverage Demonstrating Proof-of-Possession when using OidcClient to build a native OIDC client.
sidebar:
  label: DPoP
  order: 1
---

[DPoP][dpop-spec] specifies how to bind an asymmetric key stored
within a JSON Web Key (JWK) to an access token. This will make the access token bound to the key such that if the
access token were to leak, it cannot be used without also having access to the private key of the corresponding JWK.

The `Duende.IdentityModel.OidcClient.Extensions` library adds supports for DPoP to OidcClient.

## DPoP Key

> Before we begin, your application needs to have a DPoP key, in the form of a
> JSON Web Key (or JWK). According to the [DPoP specification][dpop-spec], this
> key needs to use an asymmetric algorithm ("RS", "ES", or "PS" style).

Reviewer: ```suggestion
Before we begin, your application needs to have a DPoP key in the form of a
JSON Web Key (or JWK). According to the [DPoP specification][dpop-spec], this
key needs to use an asymmetric algorithm ("RS", "ES", or "PS" style).
```

:::note

> This means that the client application is responsible for creating the DPoP key,

Reviewer: ```suggestion
The client application is responsible for creating the DPoP key,
```

rotating it, and managing its lifetime. For as long as there are access tokens
(and possibly refresh tokens) bound to a DPoP key, that key needs to remain
available to the client application.
:::

> Creating a JWK in .NET is pretty straightforward using the `Duende.IdentityModel.OidcClient.Extensions` library:

Reviewer: ```suggestion
You can create a JWK in .NET using the `Duende.IdentityModel.OidcClient.Extensions` library.
The `JsonWebKeys` class has several static methods to help with creating JWKs using various algorithms.
```

Nothing is straightforward or simple :)

```csharp
// Program.cs
using Duende.IdentityModel.OidcClient.DPoP;

// Creates a JWK using the PS256 algorithm:
var jwk = JsonWebKeys.CreateRsaJson();

Console.WriteLine(jwk);
```

:::caution
In a production scenario, you'll want to store this JWK in a secure location

> and use DataProtection to further protect the JWK.

Reviewer: ```suggestion
and use data protection to further protect the JWK.
```

Also consider adding a link to data protection docs on our end?

:::

## Initializing the OIDC client with DPoP support

> Now that we have a JWK, we need to extend the `OidcClientOptions` to configure DPoP:

Reviewer: Consider adding one or two more sentences explaining the flow here ("options are created, then a JWK is created and using the ConfigureDpop method the OidcClientOptions are configured to use the JWK")

```csharp
// Program.cs
using Duende.IdentityModel.OidcClient;
using Duende.IdentityModel.OidcClient.DPoP;

var options = new OidcClientOptions
{
    Authority = "https://demo.duendesoftware.com",
    ClientId = "native.dpop",
    Scope = "openid profile email offline_access",
    // ...
};

// creates a new JWK, or returns an existing one
var jwk = GetDPoPJwk();

// Enable DPoP
options.ConfigureDPoP(jwk);

var oidcClient = new OidcClient(options);
```

## Proof Tokens for the API

After configuring the `OidcClientOptions` with DPoP support and creating an
`OidcClient` instance, you can use this instance to create an `HttpMessageHandler` which
will:

- manage access and refresh tokens
- add DPoP proof tokens to HTTP requests

> 

Reviewer: ```suggestion

The `OidcClient` provides `CreateDPoPHandler` as a convenience method to create such handler, which can be used with the .NET `HttpClient`.

```
